### PR TITLE
[Scheduling] Provide an ASAP scheduler for the basic problem.

### PR DESCRIPTION
**Pull request by jopperm**

### CODE REVIEW

**Comment by teqdruid** [comment]
This is the point where you actually have the ability to test the scheduler? If so, I would put them in this PR (despite the plan we discussed). This PR is digestible enough to include the tests as well, assuming the tests are small-ish and straight-forward-ish.

**Comment by jopperm** [comment]
> This is the point where you actually have the ability to test the scheduler?

It is, please see the new commit.

_Rationale for touching_ `circt-opt`:
I'd argue that schedulers are much closer to analyses than to conversions or transforms. This seems to be the first analysis-like test pass in CIRCT, so I drew inspiration from the [alias analysis test passes](https://github.com/llvm/llvm-project/blob/main/mlir/test/lib/Analysis/TestAliasAnalysis.cpp) in MLIR (and how they're [declared](https://github.com/llvm/llvm-project/blob/7655061cc64d76615e9215cd6dcbf5ee186363be/mlir/tools/mlir-opt/mlir-opt.cpp#L30) and [registered](https://github.com/llvm/llvm-project/blob/7655061cc64d76615e9215cd6dcbf5ee186363be/mlir/tools/mlir-opt/mlir-opt.cpp#L192) in `mlir-opt`.

**Comment by mikeurbach** [comment]
> I'd argue that schedulers are much closer to analyses than to conversions or transforms. 

I agree, and I would expect users of the Scheduling functionality may be implementing analyses themselves (this is something I may try myself). The way you've set up the tests here as test passes makes good sense to me. I personally don't have any concern about having them included in circt-opt. From what I've seen, -opt tools in MLIR tend to be catch-all things that are meant for compiler developers and testing functionality that exists in-tree, not production grade tools that care about being slim.

File: lib/Scheduling/ASAPScheduler.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
+//===- ASAPScheduler.cpp - As-soon-as-possible list scheduler -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of an as-soon-as-possible list scheduler for acyclic problems.
```
Comment (mikeurbach) [inline lib/Scheduling/ASAPScheduler.cpp:9]: I think this loops forever in the presence of a cycle? I know this is sort of a toy to demonstrate the functionality, but should it detect cycles and fail if it finds one? I think if you kept a set of seen `Operator *`s, or something like that, this wouldn't be too crazy to add.

File: lib/Scheduling/ASAPScheduler.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+using namespace circt;
+using namespace circt::scheduling;
+
+LogicalResult ASAPScheduler::schedule() {
+  if (failed(check()))
+    return failure();
+
+  // initialize a worklist with the block's operations
+  std::deque<Operation *> worklist;
```
Comment (mikeurbach) [inline lib/Scheduling/ASAPScheduler.cpp:25]: Could you use a `SmallVector` for this: https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h? I'm not sure the overheads of a `deque` are warrented (see that document's recommendations on `deque`). If not a `SmallVector`, perhaps just a `std::vector`? That's how one of the core upstream pattern drivers stores its own worklist, for example: https://github.com/llvm/llvm-project/blob/d4ed965b2d14361e4273525a9888344a25f3800c/mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp#L130

**Review by mikeurbach — COMMENTED** [review]
Some minor comments from my end, but this looks good to me.

**Comment by jopperm** [comment]
Thanks @mikeurbach! Yeah, you're right, this infinite loop situation is bit awkward. I think the proper solution is to have an actual cycle detection in the problem `check()` phase, but I haven't built that yet and it will probably become a PR on its own.

For this PR, I'll add a simple fail-safe for cyclic graphs - switching to draft until that is done.

File: lib/Scheduling/ASAPScheduler.cpp (hunk 1 above)
Comment (jopperm) [reply lib/Scheduling/ASAPScheduler.cpp:9]: Fail-safe added.

File: lib/Scheduling/ASAPScheduler.cpp (hunk 2 above)
Comment (jopperm) [reply lib/Scheduling/ASAPScheduler.cpp]: Yes, I switched to a `SmallVector`, as I no longer needed the queue behavior after implementing the fail-safe.

**Comment by jopperm** [comment]
The ASAP scheduler now fails gracefully on cyclic graphs. PR should be ready to review again.

**Review by mikeurbach — APPROVED** [review]
Thanks for reworking it, no further comments from me!

**Review by teqdruid — COMMENTED** [review]
I just skimmed this PR and read the tests. Nice! Thanks for iterating on this stuff.

**Comment by jopperm** [comment]
Thank you for taking the time to review! I guess this is ready to merge then; can one of you press the button, please?

**Comment by teqdruid** [comment]
Done! You should request contributor access via https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access .

**Pull request merged**